prefect_github/repository.py

```diff
@@ -23,6 +31,121 @@
 return_fields_defaults = initialize_return_fields_defaults(config_path)
 
 
+class GitHubRepository(ReadableDeploymentStorage):
+    """
+    Interact with files stored on GitHub repositories.
+    """
+
+    _block_type_name = "GitHub Repository"
+    _logo_url = "https://images.ctfassets.net/gm98wzqotmnx/187oCWsD18m5yooahq1vU0/ace41e99ab6dc40c53e5584365a33821/github.png?h=250"  # noqa: E501
+
+    repository: str = Field(
+        default=...,
+        description=(
+            "The URL of a GitHub repository to read from, in either HTTPS or SSH "
+            "format. If you are using a private repo, it must be in the HTTPS format."
+        ),
+    )
+    reference: Optional[str] = Field(
+        default=None,
+        description="An optional reference to pin to; can be a branch name or tag.",
+    )
+    credentials: Optional[GitHubCredentials] = Field(
+        default=None,
+        description="An optional GitHubCredentials block for using private GitHub repos.",  # noqa: E501
+    )
+
+    @validator("credentials")
+    def _ensure_credentials_go_with_https(cls, v: str, values: dict):
+        """Ensure that credentials are not provided with 'SSH' formatted GitHub URLs."""
+        if v is not None:
+            if not values["repository"].startswith("https://"):
+                raise InvalidRepositoryURLError(
+                    (
+                        "Crendentials can only be used with GitHub repositories "
+                        "using the 'HTTPS' format. You must either remove the "
+                        "credential if you wish to use the 'SSH' format and are not "
+                        "using a private repository, or you must change the repository "
+                        "url to the 'HTTPS' format. "
+                    )
+                )
+
+        return v
+
+    def _create_repo_url(self) -> str:
+        """Format the URL provided to the `git clone` command.
+
+        For private repos: https://<oauth-key>@github.com/<username>/<repo>.git
+        All other repos should be the same as `self.repository`.
+        """
+
+        if self.repository.startswith("https://") and self.credentials is not None:
+            repo_url = self.repository[8:]
+            token_value = self.credentials.token.get_secret_value()
+            full_url = f"https://{token_value}@{repo_url}"
```

Comment:
I think [`urllib`](https://docs.python.org/3/library/urllib.parse.html) could be valuable for the parsing and editing of URLs in this method

Reply:
This was helpful - I wasn't previously familiar with urllib. I figured out deconstructing the URLs pretty quickly, but I might have missed something for editing the URLs